contracts/mapping-token/contracts/v2/Backing.sol

```diff
@@ -0,0 +1,251 @@
+        _changeDailyLimit(mappingToken, amount);
+    }
+
+    function updateFeeMarket(address newFeeMarket) external onlyOwner {
+        feeMarket = newFeeMarket;
+    }
+
+    // here add InBoundLane, and remote issuing module must add the corresponding OutBoundLane
+    function addInboundLane(address mappingTokenFactory, address inboundLane) external onlyOwner {
+        uint32 bridgedChainPosition = IMessageVerifier(inboundLane).bridgedChainPosition();
+        require(remoteChainPosition == bridgedChainPosition, "Backing:Invalid bridged chain position");
+        uint32 bridgedLanePosition = IMessageVerifier(inboundLane).bridgedLanePosition();
+        inboundLanes[bridgedLanePosition] = inboundLane;
+        emit NewInBoundLaneAdded(mappingTokenFactory, inboundLane);
+    }
+
+    // here add OutBoundLane, and remote issuing module must add the corresponding InBoundLane
+    function addOutboundLane(address outboundLane) external onlyOwner {
+        uint32 bridgedChainPosition = IMessageVerifier(outboundLane).bridgedChainPosition();
+        require(remoteChainPosition == bridgedChainPosition, "Backing:Invalid bridged chain position");
+        uint32 bridgedLanePosition = IMessageVerifier(outboundLane).bridgedLanePosition();
+        outboundLanes[bridgedLanePosition] = outboundLane;
+        emit NewOutBoundLaneAdded(bridgedLanePosition, outboundLane);
+    }
+
+    /**
+     * @notice reigister new erc20 token to the bridge. Only owner can do this.
+     * @param bridgedLanePosition the bridged lane positon, this register message will be delived to this lane position
+     * @param token the original token address
+     * @param name the name of the original token
+     * @param symbol the symbol of the original token
+     * @param decimals the decimals of the original token
+     */
+    function registerErc20Token(
+        uint32 bridgedLanePosition,
+        address token,
+        string memory name,
+        string memory symbol,
+        uint8 decimals
+    ) external payable onlyOwner {
+        require(registeredTokens[token] == false, "Backing:token has been registered");
+
+        address outboundLane = outboundLanes[bridgedLanePosition];
+        require(outboundLane != address(0), "Backing:cannot find outboundLane to send message");
+        bytes memory newErc20Contract = abi.encodeWithSelector(
+            IMappingTokenFactory.newErc20Contract.selector,
+            address(this),
+            ERC20_TOKEN_TYPE,
+            token,
+            thisChainName,
+            name,
+            symbol,
+            decimals
+        );
+        uint256 fee = IFeeMarket(feeMarket).market_fee();
+        require(msg.value >= fee, "Backing:not enough fee to pay");
+        uint256 messageId = IOutboundLane(outboundLane).send_message{value: fee}(remoteMappingTokenFactory, newErc20Contract);
+        registerMessages[messageId] = token;
+        if (msg.value > fee) {
```

Comment:
After the discussion. we think `transfer` is simple and safe.
`transfer(wad)` == `call{value:wad,gas:2300}
Given that many places use this method, I suggest using modifier below.
```
    modifier costs() {
        uint256 fee = IFeeMarket(feeMarket).market_fee();
        if (msg.value < fee)
            revert();

        _;
        if (msg.value > fee)
            payable(msg.sender).transfer(msg.value - fee);
    }
```

Reply:
I agree if the contract is upgradable, is that ture?

This article suggest avoid using transfer since it is depend on gas cost.

If not use transfer, should use 'noreentrance + call' then.

Reply:
fixed gas cost will prevent all the fallback function execution. Use transfer, we will not care about the callback in the token contract.

Reply:
In fact we put the `transfer` at the end of the function block. It has no reentrance risk. But we think it's more security and simple if use transfer. 'noreentrance + call' is also safe, but we must introduce reentrance lib here if use this way. The mapping-token-factory is very large now. So we think `transfer` is a better choice.

Reply:
OK, looks good to me if we would like to keep simple, and do not need to consider the gas cost change for now.(This arguments from article make sense, We are suppoded to have other ways to fix that if happens)


And btw, You mean the receiver contract(not related to token contract here, right)?

Reply:
Yes. It's receiver, not token contract.

additionally, i suggest we do not support an upgradable token contracts. It's out of control when we call them.